### PR TITLE
Copy layer inputs for CNN layers

### LINKED ISSUE #2146: Can't Train a Normal CNN.
#### Issue description

When I try to train the following cnn, it only runs when I include the line 

`arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");`

However, if I comment out this line, the code no longer runs (both cases compile. but in this second case doesn't run - "Segmentation Fault"). It doesn't make sense to me why this happens.

#### Your environment

 * version of mlpack: 3.2.2
 * operating system: ubuntu 18.04
 * compiler: g++ 7.4.0
 * version of dependencies (Boost/Armadillo): libboost-1.65.1, armadillo-9.800.3

#### Steps to reproduce

 Here's the code to reproduce the issue: 

    #include <mlpack/core.hpp>
    #include <mlpack/core/data/split_data.hpp>
    #include <mlpack/methods/ann/layer/layer.hpp>
    #include <mlpack/methods/ann/ffn.hpp>
    #include <ensmallen.hpp>

    using namespace mlpack;
    using namespace mlpack::ann;
    using namespace arma;
    using namespace std;
    using namespace ens;

    int main()
    {
        double RATIO = 0.1;
        int CYCLES = 10;
        const int ITERATIONS_PER_CYCLE = 1;
        double STEP_SIZE = 10e-3;
        int BATCH_SIZE = 1;

        // Reading in image.
        cout << "Reading data ..." << endl;
        arma::Mat<unsigned char> matrix;
        data::ImageInfo info;
        data::Load("saitama.png", matrix, info, false, true);

        FFN<NegativeLogLikelihood<>, RandomInitialization> model;

        model.Add<Convolution<> >(info.Channels(), 6, 5, 5, 1, 1, 0, 0, info.Width(), info.Height());
        model.Add<ReLULayer<> >();
        model.Add<MaxPooling<> >(2, 2, 2, 2, true);
        model.Add<Linear<> >(62*62*6, 2);
        model.Add<LogSoftMax<> >();
        cout << "Training ..." << endl;

        // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
        SGD<AdamUpdate> optimizer(
          STEP_SIZE,
          BATCH_SIZE,
          ITERATIONS_PER_CYCLE,
          1e-8,
          true,
          AdamUpdate(1e-8, 0.9, 0.999));

        arma::mat trainY;
        trainY << 1;// << 0;
        arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");
        trainY.print("Train label: \n");
        cout << trainY.n_rows << " " << trainY.n_cols;

        for (int i = 1; i <= CYCLES; i++)
        {
          model.Train(arma::conv_to<arma::mat>::from(matrix), trainY, optimizer, PrintLoss(), ProgressBar());
        }
        return 0;
    }

(The image size I used is 128 x 128).

compiled using: 
```g++ cnn.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp -DHAS_STB```
#### Expected behavior

When the line is commented out, the network should still train. The code should run irrespective of whether the line in question is commented out or not.

#### Actual behavior

The code runs only when the line is NOT commented out.

### PR DESCRIPTION
**Pull request by rcurtin**
This PR is an attempt to fix #2146, although not in the way that I would have hoped.  Unfortunately, it seems that for now, for the sake of memory safety, we will need to copy the input matrix.

Specifically, the problem is this: layers like `Convolution` and `AtrousConvolution` and `TransposedConvolution` (and some others) create an alias of the given input matrix when `Forward()` is called; this is the member `inputTemp`.  This alias uses the same memory as the given input matrix; it does not make a copy, and it is not responsible for freeing the memory when the alias's destructor is called.  That member `inputTemp` may then be used in `Backward()` and `Gradient()`.

However, the problem is that there exist layer types that will cause the input matrix to other layers to be freed during the forward pass; this means that any layer that took an alias during the forward pass is now holding a pointer to invalid memory.  (I can't remember if it is individual layers that cause the freeing, or if that happens as part of the `FFN` and `RNN` classes.  In either case, the result is the same.)

I thought that perhaps we could simply use `std::move()` instead of creating an alias, but this turns out to not work because other parts of the `FFN` and `RNN` forward/backward passes rely on having the layer input still available even after `Forward()` is called on the layer.  So, for now, I've simply copied the input.  That's not the fastest solution, but I think it's better to have a slow solution that works instead of a fast one that sometimes segfaults. :)

Some points for discussion in this PR:

 - If the layer input is still expected even after `Forward()` is called by the `FFN` class, then should forward really take `arma::mat&&`s for both input and output?  Or should the input be, e.g., `const arma::mat&`?

 - Have we written anywhere the specific expectations with what can and can't be done with the inputs to `Forward()` implementations?  (I guess that's kind of related to the first question.)

 - Is it possible to refactor `FFN` and `RNN` such that the `std::move()` approach (instead of the copy approach I used here) can work?

I ask these questions because I'm not particularly familiar with the internal design of this part of the code, so maybe there is just something I'm not aware of, I don't know. :)  Relevant folks for this discussion: @zoq, @ShikharJ, @walragatver, @sreenikSS, @saksham189.  (Sorry for so many tags.  It just seems to me like there are a couple of fundamental issues here that might be worth discussing. :+1:)

### CODE REVIEW

**Comment by saksham189** [comment]
I came to the same conclusion about the `inputTemp` variable [here while testing](https://github.com/mlpack/mlpack/issues/1661#issuecomment-473517680) and I could not understand why it was happening.

Also, if we do go ahead with the solution we should make the same changes in [Mean Pooling](https://github.com/mlpack/mlpack/blob/e4ff78ed064e7e91f3bb1380502e96d3b01bfba7/src/mlpack/methods/ann/layer/mean_pooling_impl.hpp#L61) and [Max Pooling](https://github.com/mlpack/mlpack/blob/e4ff78ed064e7e91f3bb1380502e96d3b01bfba7/src/mlpack/methods/ann/layer/max_pooling_impl.hpp#L61)

**Comment by rcurtin** [comment]
Looks like Azure forgot to update the build status for the `Windows VS14 Plain` job... now it says that it both failed and passed. :)  I guess the failed status just didn't get removed.

**Comment by zoq** [comment]
> If the layer input is still expected even after Forward() is called by the FFN class, then should forward really take arma::mat&&s for both input and output? Or should the input be, e.g., const arma::mat&?

Thanks for looking into the issue, my first impression is that switching to `const arma::mat&` is probably the best solution.

**Comment by rcurtin** [comment]
@zoq agreed; do you think that that's something we should open a separate issue for?  (Also I guess we should change `output` to `arma::mat&` instead of `arma::mat&&` too?)

**Comment by rcurtin** [comment]
Looks like Azure forgot to update the build status for the `Windows VS14 Plain` job... now it says that it both failed and passed. :)  I guess the failed status just didn't get removed.

**Comment by zoq** [comment]
> @zoq agreed; do you think that that's something we should open a separate issue for? (Also I guess we should change `output` to `arma::mat&` instead of `arma::mat&&` too?)

Right, change input and output and yes we should open a seperate issue for that, or maybe I just do it myself. I guess the question is do we want to include the change, switching from `arma::mat&&` to `arma::mat&` into the next release?

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.  Even ones that I personally opened. :))

**Comment by rcurtin** [comment]
I don't think I'm going to merge this yet.  I'm doing a closer look at the convolution code and I'm wondering if `inputTemp` is even needed.  I should have some result today or tomorrow, and depending on the result of that, we can merge this.

**Comment by zoq** [comment]
Sounds good.

**Comment by rcurtin** [comment]
Closing in favor of #2326.
